drivers/flash/flash_andes_qspi.c

```diff
@@ -0,0 +1,968 @@
+
+	sys_write32(tctrl, QSPI_TCTRL(base));
+	/* Enable TX/RX FIFO interrupts */
+	sys_write32(int_msk, QSPI_INTEN(base));
+	/* write CMD register to send command*/
+	sys_write32(opcode, QSPI_CMD(base));
+	k_sem_take(&dev_data->device_sync_sem, K_FOREVER);
+exit:
+	return ret;
+}
+
+/* Everything necessary to acquire owning access to the device. */
+static void acquire_device(const struct device *dev)
+{
+	struct flash_andes_qspi_data *dev_data = dev->data;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+}
+
+/* Everything necessary to release access to the device. */
+static void release_device(const struct device *dev)
+{
+	struct flash_andes_qspi_data *dev_data = dev->data;
+
+	k_sem_give(&dev_data->sem);
+}
+
+/**
+ * @brief Wait until the flash is ready
+ *
+ * @param dev The device structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static int flash_andes_qspi_wait_until_ready(const struct device *dev)
+{
+	int ret;
+	uint8_t reg;
+
+	do {
+		ret = flash_andes_qspi_cmd_read(dev,
+		FLASH_ANDES_CMD_RDSR, &reg, 1);
+	} while (!ret && (reg & FLASH_ANDES_WIP_BIT));
+
+	return ret;
+}
+
+#if defined(CONFIG_FLASH_ANDES_QSPI_SFDP_RUNTIME) || \
+	defined(CONFIG_FLASH_JESD216_API)
+/*
+ * @brief Read content from the SFDP hierarchy
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ * @param addr The address to send
+ * @param data The buffer to store or read the value
+ * @param length The size of the buffer
+ * @return 0 on success, negative errno code otherwise
+ */
+static int read_sfdp(const struct device *const dev,
+		     off_t addr, void *data, size_t length)
+{
+	/* READ_SFDP requires a 24-bit address followed by a single
+	 * byte for a wait state.  This is effected by using 32-bit
+	 * address by shifting the 24-bit address up 8 bits.
+	 */
+	return flash_andes_qspi_access(dev, JESD216_CMD_READ_SFDP,
+				       ANDES_ACCESS_ADDRESSED,
+				       addr, data, length);
+}
+#endif /* CONFIG_FLASH_ANDES_QSPI_SFDP_RUNTIME */
+
+/**
+ * @brief Write the status register.
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ * @param sr The new value of the status register
+ *
+ * @return 0 on success or a negative error code.
+ */
+static int flash_andes_qspi_wrsr(const struct device *dev,
+				 uint8_t sr)
+{
+	int ret = flash_andes_qspi_cmd_write(dev, FLASH_ANDES_CMD_WREN);
+
+	if (ret == 0) {
+		ret = flash_andes_qspi_access(dev, FLASH_ANDES_CMD_WRSR,
+					      ANDES_ACCESS_WRITE, 0, &sr,
+					      sizeof(sr));
+		flash_andes_qspi_wait_until_ready(dev);
+	}
+
+	return ret;
+}
+
+static int flash_andes_qspi_read(const struct device *dev,
+				 off_t addr, void *dest, size_t size)
+{
+	const size_t flash_size = dev_flash_size(dev);
+	int ret;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0 || addr >= flash_size ||
+	     size > flash_size || (flash_size - addr) < size)) {
```

Comment:
I am really sorry, this is my mistake: the last condition should be `(flash_size - addr) > size` and you can drop the `size > flash_size`..

Reply:
You are right.